### PR TITLE
Add drop out and normalization to blocks

### PR DESCRIPTION
**Pull request by zacespinosa**
# Modulus Pull Request

Here's an example of how I added dropout and batch norm in the Multi_SymmetricConvNeXtBlock. I expect we're going to do a few experiments before converging on a specific normalization technique / code version. This is intended to be a reference to ensure we're implementing normalization similarly. 

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/modulus/blob/main/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.
- [x] The [CHANGELOG.md](https://github.com/NVIDIA/modulus/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/modulus/issues) is linked to this pull request.

### CODE REVIEW

File: modulus/models/dlwp_healpix_layers/healpix_blocks.py [hunk 1]
```diff
@@ -493,6 +492,62 @@ def forward(self, x):
...
+            latent_channels: int = 1,
+            out_channels: int = 1,
+            kernel_size: int = 3,
+            dilation: int = 1,
+            upscale_factor: int = 4,
+            n_layers: int = 1,
+            activation: th.nn.Module = None,
+            enable_nhwc: bool = False,
+            enable_healpixpad: bool = False,
+            batch_norm: bool = True,
```
Comment (nathanielcresswellclay) [inline modulus/models/dlwp_healpix_layers/healpix_blocks.py]: Perhaps set default false to for consistency between classes?

**Review by nathanielcresswellclay — CHANGES_REQUESTED** [review]
This should probably be rebased on `dev` before merging.

File: physicsnemo/models/dlwp_healpix_layers/healpix_blocks.py [hunk 2]
```diff
@@ -493,7 +492,6 @@ def forward(self, x):
         # return second convnext result
         return self.skip_module2(x1) + self.convblock2(x1)
 
-
 class Multi_SymmetricConvNeXtBlock(th.nn.Module):
```
Comment (pzharrington) [inline physicsnemo/models/dlwp_healpix_layers/healpix_blocks.py:499]: nit: follow naming convention and don't use a `_` in the name

Comment (pzharrington) [reply physicsnemo/models/dlwp_healpix_layers/healpix_blocks.py:499]: Ah dang, I see that has already made it's way into physicsnemo main branch :( nevermind then

File: physicsnemo/models/dlwp_healpix_layers/healpix_blocks.py [hunk 3]
```diff
@@ -18,9 +18,12 @@
 
 import torch
 import torch as th
-
 from .healpix_layers import HEALPixLayer
+from .normalization import ConditionalLayerNorm
 
+from hydra.utils import instantiate
+from omegaconf import DictConfig
+from omegaconf import OmegaConf
```
Comment (pzharrington) [inline physicsnemo/models/dlwp_healpix_layers/healpix_blocks.py]: `OmegaConf` is unused

File: physicsnemo/models/dlwp_healpix_layers/healpix_decoder.py [hunk 4]
```diff
@@ -19,6 +19,7 @@
 import torch as th
 from hydra.utils import instantiate
 from omegaconf import DictConfig
+from omegaconf import OmegaConf
```
Comment (pzharrington) [inline physicsnemo/models/dlwp_healpix_layers/healpix_decoder.py]: Same here

File: physicsnemo/models/dlwp_healpix_layers/healpix_encoder.py [hunk 5]
```diff
@@ -19,6 +19,7 @@
 import torch as th
 from hydra.utils import instantiate
 from omegaconf import DictConfig
+from omegaconf import OmegaConf
```
Comment (pzharrington) [inline physicsnemo/models/dlwp_healpix_layers/healpix_encoder.py]: Same here

File: physicsnemo/models/dlwp_healpix_layers/normalization.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+        """
+
+        # check if the layer is initialized
+        if not self.initialized:
+            self._initialze(x.device)
+
+        # normal layer norm, assumes channel-last format
+        mean = x.mean(dim=-1, keepdim=True)
+        var = x.var(dim=-1, keepdim=True, unbiased=False)
+        x_norm = (x - mean) / th.sqrt(var + self.eps)
```
Comment (pzharrington) [inline physicsnemo/models/dlwp_healpix_layers/normalization.py]: Noting for future optimizations, we should pivot to using the torch/apex layer norm with `affine=False` to do these ops for better performance.

File: physicsnemo/models/dlwp_healpix_layers/healpix_decoder.py [hunk 7]
```diff
@@ -150,7 +152,12 @@ def forward(self, inputs: Sequence) -> th.Tensor:
             if layer["upsamp"] is not None:
                 up = layer["upsamp"](x)
                 x = th.cat([up, inputs[-1 - n]], dim=self.channel_dim)
-            x = layer["conv"](x)
+            # apply the conv block, check if the layer accepts conditional inputs
```
Comment (pzharrington) [inline physicsnemo/models/dlwp_healpix_layers/healpix_decoder.py:155]: The check mentioned here doesn't really happen, right? May be safest to raise an error if the layer does not support the conditional arg being passed

File: physicsnemo/models/dlwp_healpix_layers/healpix_blocks.py [hunk 8]
```diff
@@ -596,103 +631,154 @@ def __init__(
...
+        # 3x3: latent → out
+        self.output_conv = geometry_layer(
+            layer=th.nn.Conv2d,
+            in_channels=int(latent_channels),
+            out_channels=out_channels,
+            kernel_size=kernel_size,
+            dilation=dilation,
+            enable_nhwc=enable_nhwc,
+            enable_healpixpad=enable_healpixpad,
+        )
```
Comment (pzharrington) [inline physicsnemo/models/dlwp_healpix_layers/healpix_blocks.py:726]: I understand why you had to unroll what was previously a simple `nn.Sequential` into these separate modules, to support passing the conditional arg. It's a bit of a pain, but for backwards compatibility of checkpoints we could consider collecting things into a `nn.ModuleList` that is still named `self.convblock`. I think then old checkpoints would load without issue, and the core part of forward would look something like:
```
for layer in self.convblock:
    if isinstance(layer, ConditionalLayerNorm):
        x = layer(x, conditions_cln)
    else:
        x = layer(x)
```
Thoughts?

File: physicsnemo/metrics/climate/healpix_loss.py [hunk 9]
```diff
@@ -307,3 +307,97 @@ def forward(self, prediction, target, average_channels=True):
...
+
+        # checks for dimensions 
+        if not prediction.shape[1:] == target.shape:
+            raise ValueError(f"Shape of prediction should match shape of target along non-ensemble dimensions, got {prediction.shape} and {target.shape}")
+    
+        if not prediction.shape[0] % self.n_members == 0:
+            raise ValueError(f"Shape of prediction should have ensemble dimension of size {self.n_members}, got {prediction.shape[0]}")
+        
+        crps_terms = []
+        for j in range(self.n_members):
```
Comment (akshaysubr) [inline physicsnemo/metrics/climate/healpix_loss.py]: @nathanielcresswellclay I think this can be optimized a lot by removing the for loops and using a vectorized implementation something like this
```python
pdist = th.nn.PairwiseDistance(p=1)

# Compute first two terms of the afCRPS formula (they are both the same)
diff = pdist(prediction - target.unsqueeze(0)) # [Cond, B] (add a unit ensemble dimension to target to subtract from all predictions)

# Compute the last term of the afCRPS formula (pairwise L1 distance matrix)
dist_matrix = pdist(predictions.unsqueeze(1), predictions.unsqueeze(0))  # [Cond, Cond, B] 

# afCRPS formula
crps_terms = diff.unsqueeze(0) + diff.unsqueeze(1) - self.coeff_eps * dist_matrix  # [Cond, Cond, B]

# Final afCRPS
crps = crps_terms.sum()
```

This doesn't account for channel specific weights, but you can just fold those into the predictions and targets. Just scale each channel by the loss weights for that channel before this code snippet.

Comment (nathanielcresswellclay) [reply physicsnemo/metrics/climate/healpix_loss.py]: This makes sense. Thanks

Comment (pzharrington) [reply physicsnemo/metrics/climate/healpix_loss.py]: Ended up not needing the `nn.PairwiseDistance` here, just pushed an updated version that computes the difference matrix simply by `torch.abs(predictions.unsqueeze(1) - predictions.unsqueeze(0))`.

Comment (akshaysubr) [reply physicsnemo/metrics/climate/healpix_loss.py]: @pzharrington Although this would be a vectorized implementation, one of the reasons to use `nn.PairwiseDistance` is to also fuse the reduction into the same operation so that you only ever have tensors of size `[Cond, B, F, T, C, H, W]` and never explicitly instantiate tensors of size `[Cond, Cond, B, F, T, C, H, W]`. With an ensemble dimension of 2, this is a 2x overhead, but would be much more significant if we wanted to go to larger ensemble dimensions

File: physicsnemo/metrics/climate/healpix_loss.py [hunk 10]
```diff
@@ -307,3 +307,102 @@ def forward(self, prediction, target, average_channels=True):
...
+        if not prediction.shape[0] % self.n_members == 0:
+            raise ValueError(f"Shape of prediction should have ensemble dimension of size {self.n_members}, got {prediction.shape[0]}")
+
+        # Compute first term of the afCRPS formula
+        diff_target = th.abs(prediction - target.unsqueeze(0)).mean(dim=0) # [B, F, T, C, H, W]
+
+        # Compute the last term of the afCRPS formula (pairwise L1 distance matrix)
+        # Index only the upper triangular part of the matrix for unique pairs
+        dist_matrix = th.abs(prediction.unsqueeze(1) - prediction.unsqueeze(0)) # [Cond, Cond, B, F, T, C, H, W]
+        diff_ensemble = dist_matrix[self.i, self.j] # [num_unique_pairs, B, F, T, C, H, W]
```
Comment (akshaysubr) [inline physicsnemo/metrics/climate/healpix_loss.py]: Does this create a view or does it explicitly index and coalesce the tensor? If it is the latter, it might be faster to keep the full matrix and add the 2 factor in the `self.averaging_coeff` denominator.

Comment (pzharrington) [reply physicsnemo/metrics/climate/healpix_loss.py]: It does explicitly index and coalesce. Adding an extra division by two makes sense, thanks! Honestly it's probably worth just writing a dedicated CRPS implementation for n=2 within this `__call__` since we expect to do a lot of training at that regime, then we can avoid the whole distance matrix entirely

File: physicsnemo/metrics/climate/healpix_loss.py (hunk 9 above)
Comment (pzharrington) [reply physicsnemo/metrics/climate/healpix_loss.py]: Ok, so with `PairwiseDistance` I can get a decent speedup (~35%) over my first implementation that indexes into the full nxn matrix, but for n=2 ensemble members it is obviously faster to just avoid either one and use `torch.abs()` directly with the first and second members when computing the second term. I think I'll have it so if `n=2` it will use that implementation, or swap to `PairwiseDistance` for `n>2`. Here are some numbers:
```
Direct n=2
Members=2, FWD=0.133 ms, FWD+BWD=0.423

PairWise
Members=2, FWD=0.213 ms, FWD+BWD=0.652
Members=4, FWD=0.598 ms, FWD+BWD=1.573
Members=8, FWD=1.990 ms, FWD+BWD=5.155
Members=16, FWD=7.249 ms, FWD+BWD=18.728

Sorted
Members=2, FWD=2.606 ms, FWD+BWD=2.845
Members=4, FWD=2.282 ms, FWD+BWD=2.773
Members=8, FWD=2.718 ms, FWD+BWD=3.803
Members=16, FWD=3.754 ms, FWD+BWD=6.510
```
The last group is from the physicsnemo kcrps implementation, which sorts the values along ensemble dimension which is more efficient for large `n` -- it looks like the crossover point for that is around `n=8` here.

File: physicsnemo/metrics/climate/healpix_loss.py (hunk 10 above)
Comment (pzharrington) [reply physicsnemo/metrics/climate/healpix_loss.py]: Swapped to use the direct n=2 implementation or the PairwiseDistance version for n>2

**Comment by pzharrington** [comment]
Moving this to #26 to unblock.
